content/tutorial/tutorial.md

```diff
@@ -513,63 +508,64 @@ class Board extends React.Component {
 var player = {score: 1, name: 'Jeff'};
 player.score = 2;
-// Now player is {score: 2, name: 'Jeff'}
+// قيمة الـ player الان : {score: 2, name: 'Jeff'}
 ```
 
-#### Data Change without Mutation {#data-change-without-mutation}
+#### تغيير البيانات بدون تعديلها بشكل مباشر {#data-change-without-mutation}
 ```javascript
 var player = {score: 1, name: 'Jeff'};
 
 var newPlayer = Object.assign({}, player, {score: 2});
-// Now player is unchanged, but newPlayer is {score: 2, name: 'Jeff'}
+// الآن يبقى player بدون تغيير
+// ولكن تكون قيمة newPlayer هي {score: 2, name: 'Jeff'}
 
-// Or if you are using object spread syntax proposal, you can write:
+/// أو إن كنت تستخدم اقتراح صياغة نشر الكائنات تستطيع كتابة
 // var newPlayer = {...player, score: 2};
 ```
 
-The end result is the same but by not mutating (or changing the underlying data) directly, we gain several benefits described below.
+النتيجة النهائيّة هي نفسها ولكن نكسب عدّة فوائد عن طريق عدم تغيير البيانات بشكل مباشر، سنذكر هذه الفوائد الآن.
 
-#### Complex Features Become Simple {#complex-features-become-simple}
+#### تبسيط الميزات المعقدة {#complex-features-become-simple}
 
-Immutability makes complex features much easier to implement. Later in this tutorial, we will implement a "time travel" feature that allows us to review the tic-tac-toe game's history and "jump back" to previous moves. This functionality isn't specific to games -- an ability to undo and redo certain actions is a common requirement in applications. Avoiding direct data mutation lets us keep previous versions of the game's history intact, and reuse them later.
+تجعل عدم قابلية التغيير من الأسهل تنفيذ الميزات المعقدة. لاحقًا في هذا الدليل سنُنفِّذ ميزة السفر عبر الزمن والتي تسمح لنا بمراجعة تاريخ الحركات السابقة في لعبة إكس-أو مع إمكانية القفز إلى الحركات السابقة. هذه الميزة ليست خاصة بالألعاب، فالقدرة على التراجع والعودة عن أفعال محددة هي متطلب شائع في التطبيقات. يسمح لك تجنّب التعديل المباشر للبيانات بالاحتفاظ بإصدارات من تاريخ تحركات اللعبة وإعادة استخدامها لاحقًا.
 
-#### Detecting Changes {#detecting-changes}
+#### كشف التغيرات {#detecting-changes}
 
-Detecting changes in mutable objects is difficult because they are modified directly. This detection requires the mutable object to be compared to previous copies of itself and the entire object tree to be traversed.
+يصعب كشف التغيّرات في الكائنات القابلة للتعديل لأنّها تُعدَّل بشكل مباشر. يتطلب هذا الكشف مقارنة الكائنات القابلة للتعديل مع النسخ السابقة منها وكامل شجرة الكائنات المطلوبة.
 
-Detecting changes in immutable objects is considerably easier. If the immutable object that is being referenced is different than the previous one, then the object has changed.
+يُعتبَر كشف التغيّرات في الكائنات غير القابلة للتعديل أسهل. إن كان الكائن غير القابل للتعديل مختلفًا عن السابق فنعتبر أنّ الكائن قد تغيّر.
 
-#### Determining When to Re-Render in React {#determining-when-to-re-render-in-react}
+#### تحديد وقت إعادة التصيير في React {#determining-when-to-re-render-in-react}
```

Comment:
For me, the "when to re-render" expression is related to time=>timers which is not the case for us. I believe that using `متى` is better since it refers to "reaction to an event".

```suggestion
#### تحديد متى يتم إعادة التصيير في React {#determining-when-to-re-render-in-react}
```

Or we should find a better expression

Reply:
`تحديد زمن  إعادة التصيير في React` !!

Reply:
<div dir="rtl">
نفس الشيء، إعادة التصيير ليست متعلقة بالزمن/وقت وإنما بالأحداث على العموم، فهل يتم ذلك بإحدى توابع المؤقتات أم بواسطة shouldComponentUpdate.

نقول، "I sleep when I get tired" بمعنى "أنام حين أتعب" أو "أنام متى شعرت بالتعب" و عليه موعد النوم في هذا المثال مُحدد بحدث وليس زمن مُعيّن.
</div>

Reply:
لا تجوز المقارنة اللغوية بين اللغة الإنجليزية و العربية .. و اظن انك تعلم هذا جيدا ... 

برغم من عدم إقتناعي الشخصي بتغير بعض الكلمات لكن لاباس بهذا تحب إطار اوبن مايند 🤖 👽 ههه ..